rustls/src/client/hs.rs

```diff
@@ -352,10 +352,18 @@ fn emit_client_hello_for_retry(
         exts.push(ClientExtension::PresharedKeyModes(psk_modes));
     }
 
-    if !config.alpn_protocols.is_empty() {
+    // Use connection-level ALPN protocols if available, otherwise use config-level protocols
+    let alpn_protocols = if !cx.data.alpn_protocols.is_empty() {
+        &cx.data.alpn_protocols
+    } else if !config.alpn_protocols.is_empty() {
+        &config.alpn_protocols
+    } else {
+        &Vec::new()
+    };
+
+    if !alpn_protocols.is_empty() {
         exts.push(ClientExtension::Protocols(Vec::from_slices(
-            &config
-                .alpn_protocols
+            &alpn_protocols
```

Comment:
We have two other mentions of `config.alpn_protocols` in this file which need to be changed to match.